rule-engine/rule-engine-api/src/main/java/org/thingsboard/rule/engine/api/util/TbNodeUtils.java

```diff
@@ -105,6 +113,37 @@ public static String processTemplate(String template, Map<String, String> data)
         return result;
     }
 
+    public static void validateEntityId(TbContext ctx, EntityId entityId) {
+        validateEntityId(ctx, entityId, String.format("Entity %s [%s] does not belong to the tenant",
+                entityId.getEntityType(), entityId));
+    }
+
+    public static void validateEntityId(TbContext ctx, EntityId entityId, String msg) {
+        try {
+            if (!ctx.getEntityService().existsByTenantIdAndId(ctx.getTenantId(), entityId)) {
+                throw new RuntimeException(msg);
+            }
+        } catch (BadSqlGrammarException e) {
+            log.error("Entity {} [{}]", entityId.getEntityType(), entityId, e);
+        }
+    }
+
+    public static void validateQueueId(TbContext ctx, QueueId queueId) {
+        if (queueId != null) {
+            boolean isRelatedToCurrentTenant = ctx.getEntityService().existsByTenantIdAndId(ctx.getTenantId(), queueId);
+            if (!isRelatedToCurrentTenant) {
+                boolean isDefaultQueue = ctx.getEntityService().existsByTenantIdAndId(TenantId.SYS_TENANT_ID, queueId);
```

Comment:
Fetching TenantProfile from cache is cheap. You should do this first and then execute ctx.getEntityService().existsByTenantIdAndId with the correct tenant. No need to execute 2 DB calls in the worst-case scenario. Please perform this optimization and think about the performance in future releases.